lib/icinga/dependency-group.cpp

```diff
@@ -0,0 +1,362 @@
+ *
+ * In case there is already an identical dependency group in the registry, the provided dependency group is merged
+ * with the existing one, and that group is returned. Otherwise, the provided dependency group is registered as is,
+ * and it's returned.
+ *
+ * @param dependencyGroup The dependency group to register.
+ */
+DependencyGroup::Ptr DependencyGroup::Register(const DependencyGroup::Ptr& dependencyGroup)
+{
+	std::lock_guard lock(m_RegistryMutex);
+	if (auto [it, inserted] = m_Registry.insert(dependencyGroup); !inserted) {
+		dependencyGroup->MoveMembersTo(*it);
+		return *it;
+	}
+	return dependencyGroup;
+}
+
+/**
+ * Detach the provided child Checkable from the specified dependency group.
+ *
+ * Unregisters all the dependency objects the child Checkable depends on from the provided dependency group and
+ * removes the dependency group from the global registry if it becomes empty afterward.
+ *
+ * @param dependencyGroup The dependency group to unregister the child Checkable from.
+ * @param child The child Checkable to detach from the dependency group.
+ *
+ * @return - Returns the dependency objects of the child Checkable that were member of the provided dependency group.
+ */
+std::set<Dependency::Ptr> DependencyGroup::Unregister(const DependencyGroup::Ptr& dependencyGroup, const Checkable::Ptr& child)
+{
+	std::lock_guard lock(m_RegistryMutex);
+	std::vector<Dependency::Ptr> dependencies;
+	if (auto it(m_Registry.find(dependencyGroup)); it != m_Registry.end()) {
+		auto existingGroup(*it);
+		dependencies = existingGroup->GetMembers(child.get());
+		if (existingGroup->GetMemberCount() == dependencies.size()) {
+			m_Registry.erase(it);
+		}
+
+		for (const auto& dependency : dependencies) {
+			existingGroup->RemoveMember(dependency);
+		}
+	}
+	return {dependencies.begin(), dependencies.end()};
+}
+
+/**
+ * Retrieve the size of the global dependency group registry.
+ *
+ * @return size_t - Returns the size of the global dependency groups registry.
+ */
+size_t DependencyGroup::GetRegistrySize()
+{
+	std::lock_guard lock(m_RegistryMutex);
+	return m_Registry.size();
+}
+
+DependencyGroup::DependencyGroup(String name, const Dependency::Ptr& dependency)
+	: DependencyGroup{std::move(name), std::set{dependency}}
+{
+}
+
+DependencyGroup::DependencyGroup(String name, const std::set<Dependency::Ptr>& dependencies): m_Name(std::move(name))
+{
+	for (const auto& dependency : dependencies) {
+		AddMember(dependency);
+	}
+}
+
+/**
+ * Create a composite key for the provided dependency.
+ *
+ * @param dependency The dependency object to create a composite key for.
+ *
+ * @return - Returns the composite key for the provided dependency.
+ */
+DependencyGroup::CompositeKeyType DependencyGroup::MakeCompositeKeyFor(const Dependency::Ptr& dependency)
+{
+	return std::make_tuple(
+		dependency->GetParent()->GetName(),
+		dependency->GetPeriodRaw(),
+		dependency->GetStateFilter(),
+		dependency->GetIgnoreSoftStates()
+	);
```

Comment:
Is there a reason for using `String` types for the first to instead of pointers to the corresponding objects? Pointers need less memory and are faster to compare.